src/MicrosoftTeams.ts

```diff
@@ -564,23 +736,64 @@ namespace microsoftTeams
             return;
         }
 
-        // Process only if the message is coming from a valid origin or the origin used for local testing
+        // Process only if the message is coming from a valid origin
+        let messageSource = evt.source || evt.originalEvent.source;
         let messageOrigin = evt.origin || evt.originalEvent.origin;
-        if (validOrigins.indexOf(messageOrigin.toLowerCase()) === -1)
+        if ((validOrigins.indexOf(messageOrigin.toLowerCase()) === -1) &&
+            (messageOrigin !== currentWindow.location.origin) ||
+            (messageSource === currentWindow))
         {
             return;
         }
 
-        // Set our parent origin so that we can use it when sending messages
-        parentOrigin = messageOrigin;
+        // Update our parent and child relationships based on this message
+        updateRelationships(messageSource, messageOrigin);
 
-        // If we have any messages in our queue send them now
-        while (messageQueue.length > 0)
+        // Handle the message
+        if (messageSource === parentWindow)
+        {
+            handleParentMessage(evt);
+        }
+        else if (messageSource === childWindow)
         {
-            parentWindow.postMessage(messageQueue.shift(), parentOrigin);
+            handleChildMessage(evt);
         }
+    }
 
-        // Check to see if this looks like a request or a response
+    function updateRelationships(messageSource: Window, messageOrigin: string): void
+    {
+        // Determine whether the source of the message is our parent or child and update our
+        // window and origin pointer accordingly
+        if (!parentWindow || (messageSource === parentWindow))
+        {
+            parentWindow = messageSource;
+            parentOrigin = messageOrigin;
+        }
+        else if (!childWindow || (messageSource === childWindow))
+        {
+            childWindow = messageSource;
+            childOrigin = messageOrigin;
+        }
+
```

Comment:
This is interesting. If the parent or child sites are compromised, we'll still listen to them regardless with the next message they send us. navigateCrossDomain seems only necessary if the parent is the client? I'm not sure if this is strong enough security.

If the parent is the client, parentOrigin should never change. If the parent is not the client and the parent navigates cross origin (telling its parent), without this communication would break.

If the child navigates cross origin, they would call navigateCrossDomain which the parent can handle just as the client does.

So the non-client cross origin navigate while the popup is up is the only case that might not work.

Thoughts?